add a new blog post on a data analysis workflow with git-worktrees and datalad (#7)
datalad/blog: content/posts/git-worktree-workflow/index.md

Today's solutions often fall into two extremes: heavyweight reproducible workflows like [fairly-big-processing-workflow](https://github.com/psychoinformatics-de/fairly-big-processing-workflow) that are beautiful for final analysis but feel like overkill for daily development, or ad-hoc scripts that get the job done but leave no trace of how they did it.

At the recent [Distribits](https://www.distribits.live/events/2025-distribits/) meeting, I shared my struggle with [Yarik](https://github.com/yarikoptic), and confessed that I never fully understood the [YODA principle](https://handbook.datalad.org/en/latest/basics/101-127-yoda.html) - I was on the way, but I couldn't see clearly where I was heading. Together, we've found a sweet spot that gives me both **fast iteration during development** and **clean reproducibility for batch processing** - all without duplicating data or rebuilding containers. The secret? [Git worktrees](https://git-scm.com/docs/git-worktree) combined with [DataLad](https://handbook.datalad.org/en/latest/index.html)'s nested datasets.

> ### When YODA's wisdom grows on trees

Reviewer: I felt that the jump to details  is a bit too sudden for those who aren't familiar with yoda. Might be worth adding a sentence or two on work trees and composition of datasets via submodules here as a principle of modularity in Yoda principles

Reviewer: May be even pointing to other examples like datasets in OpenNeuroDerivatives to reattest that it is somewhat "common" and not that scary

Author: Ok, I'll add a more genuine introduction on the topic - YODA or maybe even better - on principles of organizing a project.
I have no idea what you mean by OpenNeuroDerivatives... Maybe it's not that "common" :P

Reviewer: see the ones under https://github.com/OpenNeuroDerivatives/ . Another potentially useful is https://github.com/OpenNeuroStudies/ which is aggregating together derivatives with original datasets

> ```

Reviewer: Could you also add a sentence or two on the opening on what your study is about and what you are aiming here at (preprocessing vs eg paper figures)

Meng_Psychedelics/ (superdataset)
├── raw/
│ ├── L5b_2p/ (subdataset)
│ └── L5b_bpod/ (subdataset)
├── derived/
│ └── L5b/ (subdataset)
│   ├── inputs/
│   │ ├── L5b_2p/ (subdataset → ../../raw/L5b_2p)
│   │ └── L5b_bpod/ (subdataset → ../../raw/L5b_bpod)
│   ├── code/ (subdataset)
│   ├── containers/ (subdataset)
│   ├── 01_suite2p/
│   ├── ...
│   └── 04_dataframes/
└── ...
```

> The active work in my project happens in the 'derived/L5b' subdataset that consumes raw data as inputs and produces multiple intermediate outputs, e.g. '01_suite2p' ... '04_dataframes'. The subdataset 'code' is a pure git repo and I use [Jujutsu](https://docs.jj-vcs.dev/latest/) for active development (because it's such a beautiful tool). After a rather nerve-wracking experience of mixing datalad with jj - it was like Schrödinger's cat, everything was simultaneously staged and unstaged 😱 - I restrict jj usage to 'code', and continue using datalad for managing all subdatasets with annexed content, orchestrating across nested datasets as a whole, and for captureing provenance of all derived data and figures with the datalad (containers-) run command.

Reviewer: This is such a wonderful example of mixing different tech (git annex and datalad , with jj) which is built on top of the same (git) foundation using that foundation basic constructs (repos, commits, submodules) ... I wonder if we could add some kind of message here too for people to use such core text instead of creating new and ad hoc stuff

Author: That's a very complex and thought-provoking comment! I think you are saying: exploit more of the core, well-designed foundational components such as git's native features, and don't just layer tools upon tools that are all based on the same foundation (git) anyway. 

So here is the thing: jj is architecturally different. It's not a git extension but a separate VCS with git compatibility. It has its own storage backend and data model (e.g., no staging area, working copy as commit, primary objects are changes and not commits). Not only that, but it feels really different to me as a user than git or datalad. It unblocks me, and makes me loose fear of experimenting. I think people are attracted to jj because it is so extremely well designed. It's like switching from a conventional keyboard to an ergonomic keyboard -- I'm suddenly relieved from back pains because having the hands and elbows properly positioned, I sit straight. 

So my 'nerve-wracking experience' wasn't about mixing layers of Git-based tools — it was about integrating two different version control systems (jj ↔ datalad-on-git). Of course, it is not the standard way of doing things, and I acknowledge the "ad hoc"-ness of my solution, but I'd love to find a way to extend this discussion and maybe improve this ad hoc stuff I'm writing about. It's actually really fun and feels powerful! Maybe you would like to read about [my experience with jj and datalad](https://gin.g-node.org/jwu/Meng_Psychedelics/wiki/Unlocking+martial+magic+with+Datalad+and+Jujutsu) I jotted down a few weeks ago.

Reviewer: I chatted my reply to you interactively!

> The active work in my project happens in the 'derived/L5b' subdataset that consumes raw data as inputs and produces multiple intermediate outputs, e.g. '01_suite2p' ... '04_dataframes'. The subdataset 'code' is a pure git repo and I use [Jujutsu](https://docs.jj-vcs.dev/latest/) for active development (because it's such a beautiful tool). After a rather nerve-wracking experience of mixing datalad with jj - it was like Schrödinger's cat, everything was simultaneously staged and unstaged 😱 - I restrict jj usage to 'code', and continue using datalad for managing all subdatasets with annexed content, orchestrating across nested datasets as a whole, and for captureing provenance of all derived data and figures with the datalad (containers-) run command.

Reviewer: ```suggestion
The active work in my project happens in the 'derived/L5b' subdataset that consumes raw data as inputs and produces multiple intermediate outputs, e.g. '01_suite2p' ... '04_dataframes'. The subdataset 'code' is a pure git repo and I use [Jujutsu](https://docs.jj-vcs.dev/latest/) for active development (because it's such a beautiful tool). After a rather nerve-wracking experience of mixing datalad with `jj` - it was like Schrödinger's cat, everything was simultaneously staged and unstaged 😱 - I restrict `jj` usage to 'code', and continue using `datalad` for managing all subdatasets with annexed content, orchestrating across nested datasets as a whole, and for capturing provenance of all derived data and figures with the [datalad (containers-)](https://docs.datalad.org/projects/container/en/latest/generated/man/datalad-containers-run.html) [run](https://docs.datalad.org/en/latest/generated/man/datalad-run.html#datalad-run) command.
```

just to 
- accent where it is a CLI command (it is DataLad project , and `datalad` command), same for `jj`
- immediately give urls to folks to see what those "run" command about

### Neither ready to go big nor go home

> Usually, I craft one datalad run command, test it, and then write a bash script to run it for multiple subjects and experiments. Here is when it becomes painful: once it's running you can't carry on working, because any modification will interrupt the next datalad run command. Ideally, I can run things in an isolated environment that is clean and stays untouched until the job finishes. Git worktrees come in handy, because they essentially access the same annexed content (so no time wasted in getting content for another clone), but live in a different directory on the disk. Worktrees share the annex and repo history, but not the state the repo is in. This allows me to checkout different branches simultaneously since they live on different worktrees.

Reviewer: ```suggestion
Usually, I craft one [`datalad run`](https://docs.datalad.org/en/latest/generated/man/datalad-run.html) command, test it, and then write a bash script to run it for multiple subjects and experiments. Here is when it becomes painful: once it's running you can't carry on working, because any modification will interrupt the next `datalad run` command. Ideally, I can run things in an isolated environment that is clean and stays untouched until the job finishes. Git worktrees come in handy, because they essentially access the same annexed content (so no time wasted in getting content for another clone), but live in a different directory on the disk. Worktrees share the annex and repo history, but not the state the repo is in. This allows me to checkout different branches simultaneously since they live on different worktrees.
```

> Since there is no automatic way to create a nested worktree (described in [this issue](https://github.com/datalad/datalad/issues/7764)), I manually create a worktree for each subdataset inside 'derived/L5b':

Reviewer: ```suggestion
Since at the moment there is no DataLad helper command to create a collection of nested worktrees (described in [this issue](https://github.com/datalad/datalad/issues/7764)) for each subdataset, I directly use `git worktree add` command to create a worktree for each subdataset inside 'derived/L5b':
```

A notion of "automated" is unclear here and in general there is a `git` command which already "automates" to a degree it does ;)

Reviewer: now I am thinking that in principle may be we could concoct an invocation with [`datalad foreach-dataset`](https://docs.datalad.org/en/stable/generated/man/datalad-foreach-dataset.html) to make it across all subdatasets to "automate"

Author: I see what you mean. I guess I was missing something like `git worktree add <path/to/worktree> -b <branch> -r` to do it recursively, but `-r` it's definitely something I use a lot with Datalad, and not so much with git. From a user's perspective my claim made sense, but not from a developer's perspective :) 

I thought about `datalad foreach-dataset` too! Ever since you showed me that command, I use it a lot haha. So the problem is that for each subdataset a new path has to be specified, so it's not the same for each. Or could we use some variable in that command?

```bash
# add a worktree on a new branch with the name 'runs' for the superds
M/d/L5b (master)> git worktree add /mnt/Data/L5b-runs -b runs
Preparing worktree (new branch 'runs')
HEAD is now at fde26554f [DATALAD RUNCMD] plot rois, traces and F displots for mouse J, Lisuride

# repeat for each subdataset
M/d/L5b (master)> cd inputs/L5b_bpod
M/d/L/i/L5b_bpod (master)> git worktree add /mnt/Data/L5b-runs/inputs/L5b_bpod -b runs
Preparing worktree (new branch 'runs')
HEAD is now at e32a9bbf2c [DATALAD RUNCMD] Rename bpod files of mouse J, LSD with the missing date.

M/d/L/i/L5b_bpod (master)> cd ../L5b_2p
M/d/L/i/L5b_2p (master)> git worktree add /mnt/Data/L5b-runs/inputs/L5b_2p -b runs
Preparing worktree (new branch 'runs')
HEAD is now at 3bd0b284b2 rename interrupted recording

M/d/L/i/L5b_2p (master)> cd ../../code
M/d/L/code> git worktree add /mnt/Data/L5b-runs/code runs # here I skip -b because the branch already exists
Preparing worktree (checking out 'runs')
HEAD is now at ef5dafb generate_plots: plot rois, traces, and F displots
```

> The branch for 'code' has already been created, when I got ready with the code:

Reviewer: ```suggestion
The branch 'runs' in `code/` has already been created, when I got ready with the code:
```

```bash
# the new datalad run command is ready
M/d/L/code> git switch -c runs
Switched to a new branch 'runs'

# 'runs' points to the commit ef5dafb4: generate_plots
M/d/L/code> jj log
Done importing changes from the underlying Git repo.
@  turmruqs jiameng.wu@gmail.com 2025-12-03 13:38:37 f8743c11
│  (empty) (no description set) # <- the clean state we need
│ ○  tnxopmxz jiameng.wu@gmail.com 2025-12-03 13:38:37 2cd4912a
├─╯  bar plot of mice IJK # <- this is just another dev branch
○  vlwmszww jiameng.wu@gmail.com 2025-12-03 13:21:37 master?? runs git_head() ef5dafb4
│  generate_plots: plot rois, traces, and F displots # <- target commit
```
For all perspective jj users, having the 'empty' change on top of the target commit is a necessary condition for a clean working tree and thus for datalad run to happen.

So the plan is to park the worktree 'runs' at this clean state on top of the commit 'ef5dafb4', run the script 'generate_plots.py' on multiple subjects and experiments, while continuing developing on dirty branches. Let's see if that works out!

> In the newly created worktree 'runs', I verify that it is in the right state that records the successful datalad run command:

Reviewer: ```suggestion
In the newly created worktree 'runs', I verify that it is in the right state that records the successful `datalad run` command:
```

```bash
# in worktree 'runs'
/m/D/L5b-runs (runs)> git log --oneline
fde26554f (HEAD -> runs, master) [DATALAD RUNCMD] plot rois, traces and F displots for mouse J, Lisuride
```

> Nice, I will first attempt a rerun! Oh wait, not so fast, I need to set up my python virtual environment. So I go:

Reviewer: ```suggestion
Nice, I will first attempt a rerun! Oh wait, not so fast, I need to set up my [Python virtual environment](https://docs.python.org/3/library/venv.html) using a wonderful and fast [`uv`](https://docs.astral.sh/uv/) tool. So I go:
```

`python` CLI in Python project

action summary:                                                                 
  add (ok: 28)
  get (notneeded: 3)
  run (ok: 1)
  save (notneeded: 4)
  unlock (ok: 28)

/m/D/L5b-runs (runs)> git log
fde26554f (HEAD -> runs, master) [DATALAD RUNCMD] plot rois, traces and F displots for mouse J, Lisuride
```
It completes without creating a new commit - the outputs are byte-for-byte identical to the previous run. This is where the magic happens: I'm running from a completely different directory, yet I don't need to `datalad get` anything or `datalad save` results. The annex's clever symlinks make the same data available in both worktrees, eliminating redundancy entirely! 🎩✨

Now it's time to write that bash script, and to test how going forth and back between worktrees works. To not mix things up, I will never develop in the worktree 'runs', but only update it with what ever has been developed and become available on 'master' - well, actually we don't care about branches in jj's universe but anyway you get the gist.

```bash
# This is the batch-processing script

> #!/usr/bin/env bash

Reviewer: I am pretty sure you better place shebang in the first line

```suggestion
#!/usr/bin/env bash
# This is the batch-processing script
```

# -------------------------------
# Datalad RUNCMD
# -------------------------------
run_one() {
  local subid="$1" drug="$2"

  echo "=== ${subid} / ${drug} ==="

  datalad run \
    -m "Plot rois, traces and distribution of F_corrected for mouse  ${subid}, ${drug}." \
    -i "04_dataframes/${subid}_Sim-Cre/${drug}/plane0/roiset_suite2p/F.pkl" \
    -i "01_suite2p/${subid}_Sim-Cre/${drug}/plane0/ops.npy" \
    -i "01_suite2p/${subid}_Sim-Cre/${drug}/plane0/stat.npy" \
    -o "05_visualization/${subid}_Sim-Cre/${drug}" \
    "python code/src/process2p/${script} {inputs} {outputs}"  || {
        echo "ERROR: datalad run failed for ${subid} / ${drug}, but continuing..."
	datalad save -m "run ${subid} / ${drug} failed, save and continue with the next one"
  	return 0  # Return success to continue the loop
    }
}

# -------------------------------
# Main loop
# -------------------------------
failed_jobs=()

for subid in "${subjects[@]}"; do
  for drug in "${drugs[@]}"; do
    if ! run_one "${subid}" "${drug}"; then
      failed_jobs+=("${subid}/${drug}")
      echo "Recorded failur:  ${subid} ${drug}"
    fi
  done
done

echo "All done."
```

> I have added a bash script and updated the 'README.md', but the git branch `runs` is still lagging behind and I'm in a 'HEAD detached' state:

Reviewer: ```suggestion
I have added this bash script under `pipelines/run_generate_plots.sh` path and updated the 'README.md', but the git branch `runs` is still lagging behind and I'm in a 'HEAD detached' state:
```

```bash
# back in the main worktree
/M/d/L/code> jj log
@  mxvrsvoq jiameng.wu@gmail.com 2025-12-04 19:57:42 708ab665
│  (empty) (no description set)
○  yknyuprl jiameng.wu@gmail.com 2025-12-04 19:56:49 git_head() 84312427
│  update README.md # <- another new commit
○  turmruqs jiameng.wu@gmail.com 2025-12-04 19:56:29 08065905
│  add bash script to run generate_plots.py # <- the new commit
│ ○  tnxopmxz jiameng.wu@gmail.com 2025-12-03 13:38:37 2cd4912a
├─╯  bar plot of mice IJK
○  vlwmszww jiameng.wu@gmail.com 2025-12-03 13:21:37 master?? runs ef5dafb4
│  generate_plots: plot rois, traces, and F displots # <- 'runs' is still here

/M/d/L/code> git status
HEAD detached from refs/heads/runs
nothing to commit, working tree clean
```

> Let's update that branch and sew the 'HEAD' back to 'runs'. There are multiple ways to do that, I use jj:

Reviewer: ```suggestion
Let's update that branch and sew the 'HEAD' back to 'runs'. There are multiple ways to do that, including plain `git` commands, I use jj which I am more familiar with:
```

Author: ```suggestion
Let's update that branch and sew the 'HEAD' back to 'runs'. There are multiple ways to do that, including plain `git` commands, but since I'm exploring Jujutsu's universe at the moment, I go:
```

add(ok): 05_visualization/240227J_Sim-Cre/displots/cell_01.png (file)           
add(ok): 05_visualization/240227J_Sim-Cre/displots/cell_02.png (file)           
add(ok): 05_visualization/240227J_Sim-Cre/displots/cell_03.png (file)           
add(ok): 05_visualization/240227J_Sim-Cre/displots/cell_04.png (file)           
add(ok): 05_visualization/240227J_Sim-Cre/displots/cell_06.png (file)           
add(ok): 05_visualization/240227J_Sim-Cre/displots/cell_07.png (file)           
add(ok): 05_visualization/240227J_Sim-Cre/displots/cell_08.png (file)           
add(ok): 05_visualization/240227J_Sim-Cre/displots/cell_somata.png (file)       
add(ok): 05_visualization/240227J_Sim-Cre/displots/cell_tuft.png (file)         
add(ok): 05_visualization/240227J_Sim-Cre/rois/all_cells.png (file)             
  [1 similar message has been suppressed; disable with datalad.ui.suppress-simil  [18 similar messages have been suppressed; disable with datalad.ui.suppress-similar-results=off]
action summary:                                                                 
  add (ok: 28)
  get (notneeded: 3)
  run (error: 1)
  save (notneeded: 5)
  unlock (ok: 28)
```

> What happened? - Apparently, I've lost the '.venv' directory during the hard reset which causes "ModuleNotFoundError: No module named 'process2p'". määäh! This actually speaks for the use of a container. The problem with the container is that I have to rebuild it everytime I update my code ... annoying! I guess that's the trade-off between efficiency and reproducibility. To illustrate this 'highly complex' dilemma with Deepseek's smart-ass comment in a graph:

Reviewer: technically that's not true: you can use container as your "environment" and run outside script in it. You can even bind-mount that script over some version of that script which you might have already inside the container. So you can kinda have the best of both worlds... might want to adjust "problem statement" here ;)

Author: I'm genuinely confused :D 
I suppose you mean the following is not true: "The problem with the container is that I have to rebuild it everytime I update my code". Proof it to me!

Author: I just did `git reset --hard runs` again, and did not loose the '.venv' directory. So maybe also this diagnosis is wrong? 

> Apparently, I've lost the '.venv' directory during the hard reset which causes "ModuleNotFoundError

Reviewer: I would need to reread above to give more conclusive answer, but overall, if you had managed to commit `.venv` and `reset --hard` away from that -- it would be gone. If it was untracked, it would remain untracked. If you want to clean up -- do `git clean -dfx` . In tandem with `git reset --hard` it is a "go nuclear" resetting to entirely clean state.

The beauty of this approach is its simplicity and efficiency:

- No data duplication: Worktrees share annexed files, so you don't need to `datalad get` everything again
- Parallel workflows: Develop in main worktree with Jujutsu or your favourite VCS, process in clean worktree with DataLad
- Seamless integration: Merge results back with a simple `git merge`
- Low overhead: Environment setup happens once, then persists for all runs (unless you are enthusiastic time-traveler like me :P)

This isn't a replacement for full reproducibility frameworks - for final publication runs, I still reach for containers and fully specified environments, and adapt something like [the fairly big processing workflow](https://github.com/psychoinformatics-de/fairly-big-processing-workflow). But for the 90% of my work that happens between initial exploration and final publication, this workflow hits the perfect balance.

Here is the destilled recipe for the workflow:
1. Create a nested git worktree for your nested dataset on a new branch, here 'runs'
2. Set up the computational environment for this worktree

> 3. Craft and test the datalad run command

Reviewer: ```suggestion
3. Craft and test the `datalad run` command invocation
```

---
title: "Git Worktrees + DataLad: The Missing Link Between Daily Development and Reproducible Analysis"
date: 2025-12-08T19:42:00+02:00
author:
- Jiameng Wu
tags:

> - Datalad

Reviewer: ```suggestion
- DataLad
```

Is there a third level of modularity? I would argue yes! If the first level is components (raw data, code, etc.) and the second is provenance (tracking how components transform into one another), then the third level is orchestration - how we safely and efficiently execute these transformations, especially when they need to happen in parallel, in isolation, or across different environments. In our labyrinth metaphor, we now have:
1. Nodes = **Components** (data, code, containers)
2. Nodes + Edges = **Provenance** (how nodes transform)
3. Paths = **Execution strategies** (how we navigate the graph)

This third level addresses questions like:
- How do I run 100 analyses in parallel? -> see this heavyweight but beautiful [fairly-big-processing-workflow](https://github.com/psychoinformatics-de/fairly-big-processing-workflow)

> - How do I ensure the same analysis runs identically on my laptop, the cluster, and in 6 months where some packages might have updated? -> @yarikoptic can you add a reference for containers here?

Reviewer: ```suggestion
- How do I ensure the same analysis runs identically on my laptop, the cluster, and in 6 months where some packages might have updated? -> see [DataLad Handbook: 7.2. Computational reproducibility with software containers](https://handbook.datalad.org/en/latest/basics/101-133-containersrun.html)
```
